### ISSUE TITLE
model: automl: Hyperparameter tuning

### ISSUE THREAD

**Issue by pdxjohnny** [intel/dffml#968]
We want to add proper AutoML to DFFML.

The goal here is to create a model which takes as a property of it's config a set of models to used for hyperparameter tuning. Another property of it's config is the set of models which we should attempt to tune (via the first set).

Default values for these results in using all installed models to try to tune all installed model plugins.

To start, we should define a reduced set of models (not all the ones we have). We'll implement automl supporting only this reduced set. The first phase of this project will be to make sure that at one model can be used to tune hyper parameters of another model.

The next phase will be to tune two models using the same tuning model.

The next phase will be to tune two models, using two models. This amounts to doing the previous phase twice, with a different tuning model the second time.

The next phase will be to go through each model in each model plugin we have and see which ones have issues being tuned using the approach taken in the previous phase. This phase will help us determine which properties or methods we may need to add to models to help them self identify and thereby indicate their requirements for hyper parameter tuning, or maybe their inherent lack of support for it.

The final phase will be to implement hyper parameter tuning for N by N models, after implementing what we found to be gaps in the previous phase.

Due to the shortened GSoC cycle, we may end up not doing all of these phases. Which one we go to will be decided as we approach the selection process.

References:
- https://scikit-learn.org/stable/model_selection.html#model-selection

Future work
- We'll later want to expand this to automate dataset cleanup. This will be more challenging and requires that we have more mature dataset cleanup capabilities implemented first. See

**Comment by BlitzCoder1** [comment]
hey i want to contribute to this can you provide a potential start as in how should i proceed

**Comment by rajpratyush** [comment]
@pdxjohnny I have experience of python and have some working experience of Automl as well (did a guided project on coursera - Automl for computer vision) Furthermore  I am currently a Kaggle contributor as well. can you provide some insights on how to begin working on this issue?

**Comment by pdxjohnny** [comment]
@TejasKarkera10 @rajpratyush Updated description

**Comment by rajpratyush** [comment]
@pdxjohnny we can begin this phase by some basic linear regression followed by ridge and lasso regression then move to other type of algorithems and i would generally go with supervised learning algorithms first

**Comment by pdxjohnny** [comment]
@rajpratyush How you choose to begin is up to you :)

Experimentation with approaches here would be wise. As it would create content which would inform ones proposal.

This issue will remain open for everyone to attempt until the GSoC proposal milestone.

**Comment by rajpratyush** [comment]
@pdxjohnny I would like to take it up as a project under GSoC 2021.

**Comment by pdxjohnny** [comment]
@rajpratyush Anyone may submit a proposal asking to work on this project during GSoC 2021.

See
- https://python-gsoc.org/students.html
- https://summerofcode.withgoogle.com/how-it-works/
- https://intel.github.io/dffml/master/contributing/gsoc/rubric.html
- https://github.com/intel/dffml/milestones?direction=desc&sort=due_date&state=open

**Comment by rajpratyush** [comment]
Ok sir @pdxjohnny btw i had asked some issues to work on can you assign them?

**Comment by pdxjohnny** [comment]
@rajpratyush Unfortunately the way GitHub org permissions work there is no way for me to use the issue assignment functionality to assign issues users who are not a member of the org a repo is in, or are not added as collaborators. I also do not have permissions to add users as collaborators. But you've read the contributing guidelines since your last comment so you know how we communicate so as to work around those issues.

**Comment by rajpratyush** [comment]
@pdxjohnny yes sir I have pushed few commits and there was needing guidance of 3 more issues so as to know where to begin with

**Comment by seraphimstreets** [comment]
I'm currently working on this. As of date, I have working plugin that is able to perform AutoML regression/classification with H2O. depending on the task specification. However, the main problem I am encountering is that H2O's Python API does not allow one to save models with a custom name, and instead saves it to the model_id. This is problematic because it means the model name is non-deterministic, and users will have to search for the exact model filename when calling the 'predict' CLI command to ensure the model is loaded. Would this be acceptable, or should I try a workaround? (I have a hack that takes the most recently modified file, but I feel that this is ugly and space-inefficient)

**Comment by pdxjohnny** [comment]
It would be good to look at https://github.com/AutoViML/featurewiz when exploring this project
